beacon-chain/rpc/prysm/v1alpha1/validator/proposer_attestations.go

```diff
@@ -287,3 +295,92 @@ func (vs *Server) deleteAttsInPool(ctx context.Context, atts []ethpb.Att) error
 	}
 	return nil
 }
+
+// filterAttestationBySignature filters attestations based on specific conditions:
+// 1. If the attestation is from the current epoch and the target is the same target, skip signature verification.
+// 2. If the attestation is from the previous epoch and the target root is the same target, skip signature verification.
+// The rest of the attestations are sent for batch signature verification. If the batch verification fails,
+// each signature is verified individually to filter out invalid attestations from the batch.
+func (vs *Server) filterAttestationBySignature(ctx context.Context, atts proposerAtts, st state.BeaconState) (proposerAtts, error) {
+	headSlot := vs.HeadFetcher.HeadSlot()
+	targetEpoch := slots.ToEpoch(headSlot)
+	headRoot, err := vs.HeadFetcher.HeadRoot(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	targetRoot, err := vs.HeadFetcher.TargetRootForEpoch([32]byte(headRoot), targetEpoch)
+	if err != nil {
+		return nil, err
+	}
+
+	currentSlot := vs.TimeFetcher.CurrentSlot()
+	currentEpoch := slots.ToEpoch(currentSlot)
+
+	var filteredAtts proposerAtts
+	for _, att := range atts {
+		attEpoch := slots.ToEpoch(att.GetData().Slot)
```

Comment:
Maybe @potuz can also verify for me, but it should be enough to verify that your target + attestation's target checkpoint match. We shouldn't need the attestation epoch.